oggm/core/preprocessing/climate.py

```diff
@@ -136,6 +136,230 @@ def process_custom_climate_data(gdir):
+
+    # get climatology data
+    loc_hgt = ncclim.get_vardata('elev')
+    loc_tmp = ncclim.get_vardata('temp')
+    loc_pre = ncclim.get_vardata('prcp')
+    loc_lon = ncclim.get_vardata('lon')
+    loc_lat = ncclim.get_vardata('lat')
+
+    # see if the center is ok
+    if not np.isfinite(loc_hgt[1, 1]):
+        # take another candidate where finite
+        isok = np.isfinite(loc_hgt)
+
+        # wait: some areas are entirely NaNs, make the subset larger
+        _margin = 1
+        while not np.any(isok):
+            _margin += 1
+            ncclim.set_subset(corners=((lon, lat), (lon, lat)), margin=_margin)
+            loc_hgt = ncclim.get_vardata('elev')
+            isok = np.isfinite(loc_hgt)
+        if _margin > 1:
+            log.debug('%s: I had to look up for far climate pixels: %s',
+                      gdir.rgi_id, _margin)
+
+        # Take the first candidate (doesn't matter which)
+        lon, lat = ncclim.grid.ll_coordinates
+        lon = lon[isok][0]
+        lat = lat[isok][0]
+        # Resubset
+        ncclim.set_subset()
+        ncclim.set_subset(corners=((lon, lat), (lon, lat)), margin=1)
+        loc_hgt = ncclim.get_vardata('elev')
+        loc_tmp = ncclim.get_vardata('temp')
+        loc_pre = ncclim.get_vardata('prcp')
+        loc_lon = ncclim.get_vardata('lon')
+        loc_lat = ncclim.get_vardata('lat')
+
+    assert np.isfinite(loc_hgt[1, 1])
+    isok = np.isfinite(loc_hgt)
+    hgt_f = loc_hgt[isok].flatten()
+    assert len(hgt_f) > 0.
+    ts_grad = np.zeros(12) + def_grad
+    if use_grad and len(hgt_f) >= 5:
+        for i in range(12):
+            loc_tmp_mth = loc_tmp[i, ...][isok].flatten()
+            slope, _, _, p_val, _ = stats.linregress(hgt_f, loc_tmp_mth)
+            ts_grad[i] = slope if (p_val < 0.01) else def_grad
+    # ... but dont exaggerate too much
+    ts_grad = np.clip(ts_grad, g_minmax[0], g_minmax[1])
+    # convert to timeserie and hydroyears
+    ts_grad = ts_grad.tolist()
+    ts_grad = ts_grad[9:] + ts_grad[0:9]
+    ts_grad = np.asarray(ts_grad * ny)
+
+    latdif=abs((temp.lat)-lat)
+    latmin=min(latdif)
+    latsel=np.where(latdif==latmin)
+    latsel=np.asarray(latsel)
+    latsel=latsel.squeeze()
+    temp = temp.isel(lat=((latsel-1), latsel, (latsel+1))) 
+    precp = precp.isel(lat=((latsel-1), latsel, (latsel+1)))
+
+    longi = np.hstack((temp.lon[:73], (temp.lon[73:]-360)))
+    temp['lon'] = longi
+    precp['lon'] = longi
+    londif = abs(temp.lon - lon)
+    lonmin=min(londif)
+    lonsel=np.where(londif==lonmin)
+    lonsel=np.asarray(lonsel)
+    lonsel=lonsel.squeeze()
+    temp=temp.isel(lon=((lonsel-1), lonsel, (lonsel+1)))
+    precp = precp.isel(lon=((lonsel-1), lonsel, (lonsel+1)))
+
+    Ndays = np.tile([31, 30, 31, 31, 28, 31, 30, 31, 30, 31, 31, 30], (y1 - y0))
+    for i in range(0,3):
+        for j in range(0,3):
+            precp[: ,i, j] = precp[: ,i, j]* (60 * 60 * 24 * 1000) * Ndays # from [m/s] to [kg/m^2 per month]
```

Comment:
No need for loops here: xarray (or numpy) could take care of the vectorized operations.